targets/TARGET_NORDIC/TARGET_NRF5x/critical_section_api.c

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2015-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hal/critical_section_api.h"
+#include "nrf_nvic.h"
+
+#include <stdint.h>
+
+static uint8_t outside_critical_region = 1;
+
+void hal_critical_section_enter()
+{
+    /* expect 1 or N calls but only lock on the first one */
+    if (outside_critical_region) {
```

Comment:
@marcuschangarm @pan- Can this implementation allow similar behavior to what I discovered before?
https://github.com/ARMmbed/mbed-os/pull/5595/files/b084812effea149c2be85b95ee675620c4371705#diff-ff24b830606200344832b47d9ac12c5c

Reply:
I think the behavior is different from what you linked. 

Here, immediately after the `outside_critical_region` check, "interrupts" are disabled. Which means that whatever thread or ISR gets to line 29 first will block all other. So a second thread will only get to execute again when the first one has exited the region. 

At that point the second thread will lock interrupts as it should.

Reply:
Looks like you are right. Thanks for explanation.